File: lib/portal-search/limits.ts
/** Default page cap for regular users. */
export const DEFAULT_MAX_PAGES = 1

/** Upper bound when an admin overrides max pages. */
export const ADMIN_MAX_PAGES = 10

export function clampPages(requested: number, isAdmin = false): number {
  const ceiling = isAdmin ? ADMIN_MAX_PAGES : DEFAULT_MAX_PAGES
  const value = Number.isFinite(requested) ? Math.trunc(requested) : DEFAULT_MAX_PAGES
  return Math.min(Math.max(value, 1), ceiling)
}

export function pagesForLimit(maxPages: number, isAdmin = false): number[] {
  const capped = clampPages(maxPages, isAdmin)
  return Array.from({ length: capped }, (_, index) => index + 1)
}

File: backend/lib/minha_casa_ai/portal_searches/limits.ex
defmodule MinhaCasaAi.PortalSearches.Limits do
  @moduledoc """
  Page and target limits for portal searches.
  Regular users are capped at one page; admins may raise the per-search limit.
  """

  @default_max_pages 1
  @admin_max_pages 10
  @max_targets 60

  def default_max_pages, do: @default_max_pages
  def admin_max_pages, do: @admin_max_pages
  def max_targets, do: @max_targets

  def max_pages(admin \\ false) do
    if admin, do: @admin_max_pages, else: @default_max_pages
  end

  def clamp_pages(requested, admin \\ false)

  def clamp_pages(requested, admin) when is_integer(requested) do
    min(max(requested, 1), max_pages(admin))
  end

  def clamp_pages(requested, admin) when is_binary(requested) do
    case Integer.parse(requested) do
      {n, _} -> clamp_pages(n, admin)
      :error -> max_pages(admin)
    end
  end

  def clamp_pages(_, admin), do: max_pages(admin)
end
